[SPARK-19668][ML] Multiple NGram sizes

mpetruska opened the pull request:
## What changes were proposed in this pull request?

[Jira ticket](https://issues.apache.org/jira/browse/SPARK-19668):
- implements extraction of multiple sizes of ngrams with `feature.NGram`

## How was this patch tested?

- unit tests were added for the implementation (`multiSliding` function)
- test cases were added to `NGramSuite` and `Word2VecSuite`

----

Benchmark of the implementation of the `multiSliding` function against a simpler, but naive implementation (tested on 1 machine):

```Scala
scala> def multiSlidingRepeated[A](x: Seq[A], min: Int, max: Int): Seq[Seq[A]] = (min to max).flatMap(x.sliding)
multiSlidingRepeated: [A](x: Seq[A], min: Int, max: Int)Seq[Seq[A]]

scala> time(multiSliding(1 to 10000, 5, 20)).toMillis
res5: Long = 66

scala> time(multiSlidingRepeated(1 to 10000, 5, 20)).toMillis
res6: Long = 434

scala> time(multiSliding(1 to 50000, 5, 20)).toMillis
res7: Long = 595

scala> time(multiSlidingRepeated(1 to 50000, 5, 20)).toMillis
java.lang.OutOfMemoryError: GC overhead limit exceeded
  ...

scala> time(multiSliding(1 to 10000, 5, 30)).toMillis
res9: Long = 91

scala> time(multiSlidingRepeated(1 to 10000, 5, 30)).toMillis
res11: Long = 428
```

mpetruska committed:
implements NGram over a range of  values

mpetruska committed:
adds tests for multiSliding

mpetruska committed:
adds additional tests for multi-length ngram interactions

mpetruska committed:
adds maxN parameter to NGram examples

mpetruska commented on mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala:
```diff
@@ -0,0 +1,121 @@
+   * )
+   * }}}
+   */
+  @Since("From which version?")
```
Please help me fill this in.

mpetruska commented on mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala:
```diff
@@ -0,0 +1,121 @@
+    }
+  }
+
+  @Since("From which version?")
```
Note to self: please don't forget to fill this in.

mpetruska commented on mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala:
```diff
@@ -0,0 +1,121 @@
+     * simultaneously.
+     * @see `org.apache.spark.ml.extensions.seq.multiSliding`
+     */
+    @Since("From which version?")
```
Note to self: please don't forget to fill this in.

AmplabJenkins commented:
Can one of the admins verify this patch?

mpetruska commented on mllib/src/main/scala/org/apache/spark/ml/feature/NGram.scala:
```diff
@@ -42,11 +42,22 @@ class NGram @Since("1.5.0") (@Since("1.5.0") override val uid: String)
+   * Default: 2, bigram features
+   * @group param
+   */
+  @Since("From which version?")
```
Note to self: please fill this in.

mpetruska commented on mllib/src/main/scala/org/apache/spark/ml/feature/NGram.scala:
```diff
@@ -57,10 +68,22 @@ class NGram @Since("1.5.0") (@Since("1.5.0") override val uid: String)
   def getN: Int = $(n)
 
+  /** @group setParam */
+  @Since("From which version?")
```
Note to self: please fill this in.

mpetruska commented on mllib/src/main/scala/org/apache/spark/ml/feature/NGram.scala:
```diff
@@ -57,10 +68,22 @@ class NGram @Since("1.5.0") (@Since("1.5.0") override val uid: String)
+  def setMaxN(value: Int): this.type = set(maxN, value)
+
+  /** @group getParam */
+  @Since("From which version?")
```
Note to self: please fill this in.

holdenk commented:
Thanks for working on this. Jenkins OK to test.

holdenk commented:
Let me CC @MLnick as well.

holdenk commented on mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala:
```diff
@@ -0,0 +1,121 @@
+   * )
+   * }}}
+   */
+  @Since("From which version?")
```
We probably don't want these APIs to be public unless there is a good reason to have them be. Maybe make them private?

holdenk replied:
Also is there a reason not to put this functionality inside of ngram its self for now?

mpetruska replied:
Right, agreed, will move the implementation into `NGram`.

holdenk commented on mllib/src/main/scala/org/apache/spark/ml/feature/NGram.scala:
```diff
@@ -42,11 +42,22 @@ class NGram @Since("1.5.0") (@Since("1.5.0") override val uid: String)
+   * @group param
+   */
+  @Since("From which version?")
+  val maxN: IntParam = new IntParam(this, "maxN", "maximum number elements per n-gram (>=n)",
```
Should this have a default value? Also perhaps the explanation that if unset maxN = n

mpetruska replied:
Added explanation in the scaladoc comment for the case when `maxN` is not set.
The default is set to 2 here: https://github.com/apache/spark/pull/19659

holdenk commented on mllib/src/test/scala/org/apache/spark/ml/feature/NGramSuite.scala:
```diff
@@ -76,11 +76,32 @@ class NGramSuite extends SparkFunSuite with MLlibTestSparkContext with DefaultRe
       .setOutputCol("myOutputCol")
       .setN(3)
+      .setMaxN(5)
     testDefaultReadWrite(t)
```
Would be good to make sure read/write makes sense even when maxN isn't set perhaps?

holdenk commented on examples/src/main/python/ml/n_gram_example.py:
```diff
@@ -35,7 +35,7 @@
     ], ["id", "words"])
 
-    ngram = NGram(n=2, inputCol="words", outputCol="ngrams")
+    ngram = NGram(n=2, maxN=4, inputCol="words", outputCol="ngrams")
```
This generally isn't enough to expose the new functionality in Python. Take a look in `feature.py` & also consider adding some Python tests.

holdenk commented on mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala:
```diff
@@ -0,0 +1,121 @@
+      }
+    }
+
+    def calculateMultiSliding(): List[B] = {
```
Does this perform as well as the current code does?

mpetruska replied:
Nope, unfortunately it's not the case:

```Scala
scala> def slidingOriginal(x: Seq[String], n: Int): Seq[String] = {
     |     x.iterator.sliding(n).withPartial(false).map(_.mkString(" ")).toSeq
     |   }
slidingOriginal: (x: Seq[String], n: Int)Seq[String]

scala> def slidingNew(x: Seq[String], n: Int): Seq[String] = {
     |     multiSliding(x, n, n).map(_.mkString(" "))
     |   }
slidingNew: (x: Seq[String], n: Int)Seq[String]

scala> time(slidingOriginal((1 to 10000).map(_.toString), 5))
res9: scala.concurrent.duration.FiniteDuration = 2427473 nanoseconds

scala> time(slidingOriginal((1 to 10000).map(_.toString), 5)).toMillis
res10: Long = 3

scala> time(slidingNew((1 to 10000).map(_.toString), 5)).toMillis
res11: Long = 15

scala> time(slidingOriginal((1 to 100000).map(_.toString), 5)).toMillis
res12: Long = 11

scala> time(slidingNew((1 to 100000).map(_.toString), 5)).toMillis
res13: Long = 479

scala> time(slidingOriginal((1 to 10000).map(_.toString), 50)).toMillis
res14: Long = 1

scala> time(slidingNew((1 to 10000).map(_.toString), 50)).toMillis
res15: Long = 136
```

Adding code to fall back to the original implementation if `n == maxN`...

holdenk reviewed (commented):
Thanks for working on this & your first PR to Spark  :) I've got some initial questions and comments, but I haven't dug too deeply. I'm sure @MLnick will have more thoughts as well :)

mpetruska committed:
moves multiSliding into NGram

mpetruska committed:
adds NGram tests for the case when maxN isn't set

mpetruska committed:
adds explanation for the case when maxN is not set in NGram

mpetruska committed:
adds special handling for the case when n == maxn in NGram

mpetruska committed:
implements maxn for pyspark ngram

mpetruska commented:
@holdenk : Thanks for the review.
Fixed/addressed your comments, can you please have another look?
Also I'd like some help with regards to the `@Since` attributes; how should I fill them in? Should I fill them in at all?

mpetruska commented:
Also, will this work in SparkR, or do I need to add some code/configuration to expose the functionality to [R]?

holdenk commented:
So SparkR will have to be separate, I'm going to cc @felixcheung , I don't know if we currently expose this functionality in the Spark R side.

As for the @Since attributes you choose them based on the version that your code is going to first appear in. At the point it might be 2.3.0 but it might also not. I'll try and do a more thorough review next week.

holdenk commented:
Is this a thing you are still working on @mpetruska ?

mpetruska commented:
No, I'm not working on this one, the feature seems to lack any real interest. We can close it, if you like.

mgaido91 commented:
@holdenk I can take this up if this is needed. Let me know. Thanks.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/14579/
Test PASSed.

holdenk commented:
I haven't seen much demand for it, so I think closing it is probably reasonable. We can re-open the issue if there is demand.

holdenk commented:
Although if you have a use case/ are interested in this @mgaido91  I'm happy to help with the review.

mgaido91 commented:
I think we can close it and reopen if someone express new interest on this, thanks @holdenk

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/15784/
Test PASSed.
